### ISSUE TITLE
`BackgroundWorker` edge cases

### ISSUE THREAD

**Issue by player-03** [openfl/lime#1502]
As it stands, [`BackgroundWorker`](https://github.com/haxelime/lime/blob/1188b3aa42b18992ef50241050217304b28c1ae5/src/lime/system/BackgroundWorker.hx) has several problems.

How it's designed to work
===================
The first problem is the lack of documentation, so let's do a quick run-through of a typical use case. For this example, suppose we have `function generateItems(type:ItemType, count:Int):Array<Item>`. It's gotten slower as the game got bigger, so we want to make it asynchronous.

```haxe
private var bgWorker:BackgroundWorker;

//Background functions must take one argument and return nothing.
private static function generateItems(params:{ type:ItemType, count:Int }):Void {
    if(params.count < 1)
        throw "Must always generate at least one item!";
    
    var items:Array<Item> = [];
    
    //Glossing over how they're actually made. Just assume it takes time.
    for(i in 0...count) {
        /* ... */
    }
    
    //Instead of returning the array, dispatch an event.
    bgWorker.sendComplete(items);
}

public function new() {
    //First things first.
    bgWorker = new BackgroundWorker();
    
    //To run a function on the other thread, add it as a `doWork` listener.
    bgWorker.doWork.add(generateItems);
    
    //Prepare the argument.
    var itemTypeAndCount = { type: ItemType.SHIELD, count: 30 };
    
    //To receive the result, add a completion listener.
    bgWorker.onComplete(function(shields:Array<Item>) {
        list = shields;
    });
    
    //Add an error listener too, just in case.
    bgWorker.onError.add(handleError);
    
    //Create the new thread and dispatch `doWork` on that thread, passing
    //the given argument to each listener.
    bgWorker.run(itemTypeAndCount);
}
```

In cases like these, `BackgroundWorker` works well.

Edge case: "completed" race conditions
==============================
Let's start with something minor. The `sendComplete()` function is run on the background thread, and sets `completed = true`. `completed` is typically checked on the main thread, which can in rare cases lead to crashes and other bugs.

```haxe
private function onEnterFrame(e:Event) {
    if(shields.bgWorker.completed)
        equipShield();
    else
        hideBehindCover();
}

private function equipShield() {
    if(!shields.bgWorker.completed)
        throw "bgWorker hasn't finished loading shields yet!";
    
    if(activeShield == null) {
        activeShield = shields.list[Std.int(Math.random() * shields.length)];
        
        equipmentWeight += activeShield.weight;
        defense += activeShield.defense;
    }
}

private function hideBehindCover() {
    if(shields.bgWorker.completed)
        throw "Shields already loaded!";
    
    if(idleFor(5)) {
        var walls = findNearbyObjects([ ObjectType.WALL, ObjectType.LOW_WALL ], 8);
        if(walls.length == 0)
            return;
        
        var destination = walls[Std.int(Math.random() * walls.length)];
        
        queueAction(Action.PATHFIND, new Vector3D(destination.x - 3, destination.y, destination.z));
        queueAction(Action.CROUCH);
    }
}
```

I admit, this code is a bit odd, but hopefully it's believable. It performs some checks for safety, but still runs into two different crashes entirely based on `sendComplete()`.

`equipShield()` is supposed to run only if `completed` is true. It checks to make sure this is the case, but it doesn't really need to. It'll always be the case. The problem is, it equates `completed` with `shields.list`, and those aren't actually the same. Here's the order things really happen in:

1. `generateItems` finishes working, and calls `sendComplete()`.
2. `sendComplete()` sets `completed` and adds the list of shields to a message queue.
3. On the main thread, `BackgroundWorker` checks the message queue, finds the message, and dispatches `onComplete`.
4. The `onComplete` listener saves the list of shields.

Problem is, step 3 happens once per update cycle. If `sendComplete()` is called at a different point in the update cycle, `completed` will be true but `shields.list` will be empty. Then `activeShield` will be null and `equipShield()` will crash.

Meanwhile, `hideBehindCover()` can crash for a completely different reason. Quite simply, it's possible that `completed` will be set after `onEnterFrame()` checks it but before `hideBehindCover()` checks it, so the assertion will fail.

The rest of `hideBehindCover()` is probably fine. I just wrote it for completeness.

Fortunately, fixing both of these is easy: don't set `complete` until the main thread receives the message.

Edge case: multiple listeners
======================
This is a more probable but less severe issue. The thing about `lime.app.Event` is you can attach as many listeners as you want. When it calls `doWork` on the other thread, it'll call each listener in turn.

To return data, normally a function would call `sendComplete()`, but that will mark the whole `BackgroundWorker` as `canceled`. (Which is counterintuitive in itself. Completion == canceling?) Once the worker has been canceled, it ignores all events from then on. So if ten functions each call `sendComplete()`, `onComplete` will be dispatched only once, and the other nine will be ignored.

This also happens with `sendError()`, but that's more to be expected.

Yes, there are some workarounds:

- Make the earlier functions pass data with `onProgress()`, and only call `onComplete()` in the last one.
- Pass an object capable of storing data to `run()`, and have the functions use this as shared storage. The final function can then call `onComplete()` to return it.

But both of these are incredibly counterintuitive. Normally you think of events are collections of unrelated (and often unordered) functions, not strictly-ordered arrays of closely-related functions. Honestly, `doWork` shouldn't be an event at all. If you really do need to call a series of functions, you can write a wrapper function for that.

Edge case: canceling and rerunning
===========================
It isn't actually possible to stop a Haxe thread from outside, so `cancel()` doesn't try to. This is too bad, because the thread will keep chugging away on whatever expensive computations it was created to perform, even though that's no longer needed. Ideally, the background thread should check the value of `canceled` so as to return early, but this isn't guaranteed.

On it's own, that isn't a problem. The background thread's events will be ignored, and eventually it will halt[.](https://xkcd.com/1266/)

Problems arise when `run()` is called again. This sets `canceled` to false and starts a new thread. However, the old thread might still be going, and now it can dispatch events again. Depending on the timing, it might finish just after the new thread starts, promptly setting `canceled` and disabling the new thread. It's similar to the "multiple listeners" edge case, only now the listeners can be executed out of order.

To fix this, we should only allow `run()` to happen once, forcing the user to create a new `BackgroundWorker` if they want another thread.

**Comment by player-03** [comment]
This didn't get closed automatically? Whoops.
